CodeEdit/Features/Settings/Pages/SourceControlSettings/Models/IgnorePatternModel.swift

```diff
@@ -0,0 +1,283 @@
+                return URL(fileURLWithPath: excludesfile) // Absolute path
+            } else {
+                return FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(excludesfile)
+            }
+        } else {
+            let defaultPath = ".gitignore_global"
+            let fileURL = FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(defaultPath)
+            await gitConfig.set(key: "core.excludesfile", value: "~/\(defaultPath)", global: true)
+            return fileURL
+        }
+    }
+
+    private func startFileMonitor() async throws {
+        let fileURL = try await gitIgnoreURL()
+        let fileDescriptor = open(fileURL.path, O_EVTONLY)
+        guard fileDescriptor != -1 else { return }
+
+        let source = DispatchSource.makeFileSystemObjectSource(
+            fileDescriptor: fileDescriptor,
+            eventMask: .write,
+            queue: DispatchQueue.main
+        )
+
+        source.setEventHandler {
+            Task {
+                await self.loadPatterns()
+            }
+        }
+
+        source.setCancelHandler {
+            close(fileDescriptor)
+        }
+
+        fileMonitor?.cancel()
+        fileMonitor = source
+        source.resume()
+    }
+
+    private func stopFileMonitor() {
+        fileMonitor?.cancel()
+        fileMonitor = nil
+    }
+
+    func loadPatterns() async {
+        await MainActor.run { loadingPatterns = true } // Ensure `loadingPatterns` is updated on the main thread
+
+        do {
+            let fileURL = try await gitIgnoreURL()
+            guard FileManager.default.fileExists(atPath: fileURL.path) else {
+                await MainActor.run {
+                    patterns = []
+                    loadingPatterns = false // Update on the main thread
+                }
+                return
+            }
+
+            if let content = try? String(contentsOf: fileURL) {
+                let parsedPatterns = content.split(separator: "\n")
+                    .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
+                    .filter { !$0.isEmpty && !$0.starts(with: "#") }
+                    .map { GlobPattern(value: String($0)) }
+
+                await MainActor.run {
+                    patterns = parsedPatterns // Update `patterns` on the main thread
+                    loadingPatterns = false  // Ensure `loadingPatterns` is updated on the main thread
+                }
+            } else {
+                await MainActor.run {
+                    patterns = []
+                    loadingPatterns = false
+                }
+            }
+        } catch {
+            print("Error loading patterns: \(error)")
+            await MainActor.run {
+                patterns = []
+                loadingPatterns = false
+            }
+        }
+    }
+
+    func getPattern(for id: UUID) -> GlobPattern? {
+        return patterns.first(where: { $0.id == id })
+    }
+
+    func savePatterns() {
```

Comment:
I dislike that this is unstructured concurrency with the loose `Task` happening. Could we add a `savingTask` variable and guard against it being nil (or cancel it) before sending this task off?

Also, we should mark the created `Task` as `@MainActor` so it keeps the correct isolation requirements for this object.

Reply:
Good idea. For your second point, is this no longer necessary now that the class is marked `@MainActor`?

Reply:
I'm not entirely certain of this, but I think that because this function is not `async`, it doesn't have the same `MainActor` requirement of the async methods on this class. I think it should be added just in case.

Reply:
Done